daemon/cmd/cni.go

```diff
@@ -0,0 +1,354 @@
+	if err != nil {
+		return nil, fmt.Errorf("could not find existing AWS CNI config for chaining %w", err)
+	}
+
+	contents, err := os.ReadFile(found)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read existing AWS CNI config %s: %w", found, err)
+	}
+
+	// We found the CNI configuration,
+	// inject Cilium as the last chained plugin
+	out, err := sjson.SetRawBytes(contents, "plugins.-1", []byte(pluginConfig))
+	if err != nil {
+		return nil, fmt.Errorf("failed to modify existing AWS CNI config at %s: %w", found, err)
+	}
+	log.Infof("Inserting cilium in to CNI configuration file at %s", found)
+	return out, nil
+}
+
+// renderCNITemplate applies any cni template replacements
+// right now, just sets .Debug and Logfile
+func renderCNITemplate(in string, opts *option.DaemonConfig) string {
+	data := struct {
+		Debug   bool
+		LogFile string
+	}{
+		Debug:   opts.Debug,
+		LogFile: opts.CNILogFile,
+	}
+
+	t := template.Must(template.New("cni").Parse(in))
+
+	out := bytes.Buffer{}
+	if err := t.Execute(&out, data); err != nil {
+		panic(err) // impossible
+	}
+	return out.String()
+}
+
+// cleanupOldCNI renames any existing CNI configuration files with the suffix
+// ".cilium_bak", excepting files in keep
+func cleanupOtherCNI(confDir, keep string) error {
+	files, err := os.ReadDir(confDir)
+	if err != nil {
+		return fmt.Errorf("failed to list CNI conf dir %s: %w", confDir, err)
+	}
+
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		name := f.Name()
+		if name == keep {
+			continue
+		}
+		if !(strings.HasSuffix(name, ".conf") || strings.HasSuffix(name, ".conflist") || strings.HasSuffix(name, ".json")) {
+			continue
+		}
+
+		log.Infof("Renaming non-Cilium CNI configuration file %s to %s.cilium_bak", name, name)
+		_ = os.Rename(path.Join(confDir, name), path.Join(confDir, name+".cilium_bak"))
+	}
+	return nil
+}
+
+// findFile looks in basedir for any filenames listed in paths.
+// returns the full path to the first one found, or error if none found
+func findFile(basedir string, paths []string) (string, error) {
+	for _, p := range paths {
+		_, err := os.Stat(path.Join(basedir, p))
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			return "", err
+		} else {
+			return path.Join(basedir, p), nil
+		}
+	}
+
+	return "", os.ErrNotExist
+}
+
+// atomicWrite writes a file atomically by first writing to a temporary file
+// then renaming it to the destination
+func atomicWrite(dir, filename string, data []byte) error {
+	tempPath := path.Join(dir, "."+filename+".tmp")
+	realPath := path.Join(dir, filename)
+
+	defer os.Remove(tempPath) // unlikely, but maybe clean this file up
+
+	if err := os.WriteFile(tempPath, data, 0644); err != nil {
+		return fmt.Errorf("failed to write temporary file %s: %w", tempPath, err)
+	}
+
+	if err := os.Rename(tempPath, realPath); err != nil {
+		return fmt.Errorf("failed to rename %s to %s: %w", tempPath, realPath, err)
+	}
+	return nil
+}
```

Comment:
If possible, let's use the `github.com/google/renameio` package for atomic writes. We already vendor it and use it in some other places in the code base.